libs/agno/agno/db/postgres/postgres.py

```diff
@@ -1143,13 +1170,221 @@ def get_metrics_table(self) -> Table:
             )
         return self.metrics_table
 
-    def get_metrics(self) -> List[AggregatedMetrics]:
-        """Get all metrics from the database."""
-        return []
+    def get_metrics_raw(
+        self, starting_date: Optional[str] = None, ending_date: Optional[str] = None, table: Optional[Table] = None
+    ) -> List[dict]:
+        try:
+            if table is None:
+                table = self.get_metrics_table()
+
+            with self.Session() as sess, sess.begin():
+                stmt = select(table)
+                if starting_date:
+                    stmt = stmt.where(table.c.date >= starting_date)
+                if ending_date:
+                    stmt = stmt.where(table.c.date <= ending_date)
+                result = sess.execute(stmt).fetchall()
+                if not result:
+                    return []
+
+                return [row._mapping for row in result]
 
-    def upsert_metrics(self) -> Optional[AggregatedMetrics]:
-        """Upsert a metrics in the database."""
-        return None
+        except Exception as e:
+            log_error(f"Exception getting metrics: {e}")
+            return []
+
+    def refresh_metrics(self) -> Optional[list[dict]]:
+        try:
+            table = self.get_metrics_table()
+
+            # Get the first day without a completed metrics record
+            with self.Session() as sess:
+                stmt = select(table).where(table.c.completed).order_by(table.c.date.desc()).limit(1)
+                result = sess.execute(stmt).fetchone()
+                if result is not None:
+                    starting_date = result._mapping["date"] + timedelta(days=1)
+                else:
+                    # No complete records found. Will start from the date of the latest incomplete record
+                    stmt = select(table).order_by(table.c.date.desc()).limit(1)
+                    result = sess.execute(stmt).fetchone()
+                    if result is not None:
+                        starting_date = result._mapping["date"]
+                    else:
+                        # No records found. Will start from the date of the first recorded session, or today
+                        first_session_date = self.get_first_session_date()
+                        starting_date = date.fromtimestamp(first_session_date) if first_session_date else date.today()
```

Comment:
Rather do one query to find the last row and continue from there? If that row is incomplete, include it in your calculations

Reply:
It should never be "today", if you can't find a metrics date, then you use the first session date as fallback

Reply:
Both points are correct! refactored